### PR TITLE
perf(local version): leverage table_id filter in read_filter to make up for removal of `CompactionGroupId`

### PR DESCRIPTION
**Pull request by soundOfDestiny**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

leverage table_id filter in read_filter to make up for removal of `CompactionGroupId`

Since we will soon be not able to use `CompactionGroupId` in filter cause group splitting, our `read_version` will contain too many SSTs from other compaction groups, over which cost too much while being iterated.
Table_id filter can solve this issue.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/storage/src/hummock/utils.rs [hunk 1]
```diff
@@ -75,7 +76,8 @@ pub fn validate_table_key_range(levels: &[Level]) -> HummockResult<()> {
     Ok(())
 }
 
-pub fn filter_single_sst<R, B>(info: &SstableInfo, key_range: &R) -> bool
+/// `table_id == 0` represents that no `table_id` filter
```
Comment (wenym1) [inline src/storage/src/hummock/utils.rs:79]: I think we will have table_id in all storage read interface. Therefore we can always get the table id before we call this method, and then we can avoid this `table_id == 0 represents that no `table_id` filter` hack. If this is only for test, we can enable this hack only in test.

File: src/storage/src/hummock/utils.rs [hunk 2]
```diff
@@ -96,7 +103,7 @@ where
     R: RangeBounds<B>,
     B: AsRef<[u8]>,
 {
-    ssts.filter(|info| filter_single_sst(info, key_range))
+    ssts.filter(|info| filter_single_sst(info, TableId::default(), key_range))
```
Comment (wenym1) [inline src/storage/src/hummock/utils.rs:106]: We can pass the table_id in the method to do better pruning.

File: src/storage/src/hummock/store/version.rs [hunk 3]
```diff
@@ -67,6 +68,7 @@ impl StagingVersion {
         &'a self,
         epoch: HummockEpoch,
         compaction_group_id: CompactionGroupId,
+        table_id: TableId,
```
Comment (wenym1) [inline src/storage/src/hummock/store/version.rs:71]: We can remove passing the compaction_group_id and only do filtering by table id.

**Review by Little-Wallace — APPROVED** [review]
LGTM

File: src/storage/src/hummock/store/version.rs (hunk 3 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/store/version.rs:71]: Does `imm` only belong to one compaction-group?

Comment (wenym1) [reply src/storage/src/hummock/store/version.rs:71]: Yes. But imm also belongs to only one table. Therefore we don't have to store the compaction group id in `imm`. table id is enough for filtering. If we don't have to pass compaction group id when we do filtering, we can remove the dependency on compaction group client on CN.

File: src/storage/src/hummock/shared_buffer/mod.rs [hunk 4]
```diff
@@ -233,7 +234,9 @@ impl SharedBuffer {
                 UncommittedData::Batch(batch) => {
                     range_overlap(key_range, batch.start_user_key(), batch.end_user_key())
                 }
-                UncommittedData::Sst((_, info)) => filter_single_sst(info, key_range),
+                UncommittedData::Sst((_, info)) => {
+                    filter_single_sst(info, TableId::default(), key_range)
```
Comment (Li0k) [inline src/storage/src/hummock/shared_buffer/mod.rs:238]: why dont we need to provide the TableID ?

File: src/storage/src/hummock/utils.rs (hunk 1 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/utils.rs]: > I think we will have table_id in all storage read interface. Therefore we can always get the table id before we call this method, and then we can avoid this `table_id == 0 represents that no `table_id` filter` hack. If this is only for test, we can enable this hack only in test.

fixed

File: src/storage/src/hummock/utils.rs (hunk 2 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/utils.rs]: > We can pass the table_id in the method to do better pruning.

fixed

File: src/storage/src/hummock/shared_buffer/mod.rs (hunk 4 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/shared_buffer/mod.rs]: > why dont we need to provide the TableID ?

fixed

File: src/storage/src/hummock/store/version.rs (hunk 3 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/store/version.rs:71]: > We can remove passing the compaction_group_id and only do filtering by table id.

fixed

File: src/storage/src/hummock/utils.rs [hunk 5]
```diff
@@ -75,28 +76,37 @@ pub fn validate_table_key_range(levels: &[Level]) -> HummockResult<()> {
...
     let table_start = user_key(table_range.left.as_slice());
     let table_end = user_key(table_range.right.as_slice());
+    #[cfg(any(test, feature = "test"))]
+    if table_id.table_id() == 0 {
+        return range_overlap(key_range, table_start, table_end);
+    }
     range_overlap(key_range, table_start, table_end)
+        && info
+            .get_table_ids()
+            .binary_search(&table_id.table_id())
```
Comment (hzxa21) [inline src/storage/src/hummock/utils.rs:94]: We assume table_ids in SstableInfo is sorted. Do we check it somewhere? If not, we should add a debug_assert in `commit_epoch` and `report_compact_task`.

**Review by hzxa21 — COMMENTED** [review]
Rest LGTM.

File: src/storage/src/hummock/shared_buffer/mod.rs [hunk 6]
```diff
@@ -233,7 +238,7 @@ impl SharedBuffer {
                 UncommittedData::Batch(batch) => {
                     range_overlap(key_range, batch.start_user_key(), batch.end_user_key())
```
Comment (wenym1) [inline src/storage/src/hummock/shared_buffer/mod.rs:239]: We may also check whether the batch.table_id matches the passed table_id

File: src/storage/src/hummock/store/version.rs [hunk 7]
```diff
@@ -67,29 +68,23 @@ impl StagingVersion {
...
     ) -> (
         impl Iterator<Item = &ImmutableMemtable> + 'a,
         impl Iterator<Item = &SstableInfo> + 'a,
     ) {
         let overlapped_imms = self.imm.iter().filter(move |imm| {
-            (compaction_group_id
-                == StaticCompactionGroupId::NewCompactionGroup as CompactionGroupId
-                || compaction_group_id == imm.compaction_group_id())
-                && imm.epoch() <= epoch
+            imm.epoch() <= epoch
```
Comment (wenym1) [inline src/storage/src/hummock/store/version.rs:78]: We may also filter by whether table_id matches.

**Review by wenym1 — COMMENTED** [review]
Rest LGTM

File: src/storage/src/hummock/shared_buffer/mod.rs (hunk 6 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/shared_buffer/mod.rs:239]: > We may also check whether the batch.table_id matches the passed table_id

I think that range overlap for batch indicates that table_id match

File: src/storage/src/hummock/store/version.rs (hunk 7 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/store/version.rs:78]: > We may also filter by whether table_id matches.

I think that range overlap for imm here indicates that table_id match

File: src/storage/src/hummock/utils.rs (hunk 5 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/utils.rs:94]: > We assume table_ids in SstableInfo is sorted. Do we check it somewhere? If not, we should add a debug_assert in `commit_epoch` and `report_compact_task`.

We didn't check it.
just added asserts in compaction

**Review by wenym1 — APPROVED** [review]

**Pull request merged**